File: islamic-kb/scripts/validate-sources.ts
import { getDatabaseConnection } from "./utils";

async function validateSources() {
  console.log("🔍 Validating Islamic Knowledge Base Integrity (TypeScript)...");
  const sql = getDatabaseConnection();

  try {
    const [surahCount] = await sql`SELECT count(*)::int FROM quran_surahs`;
    const [verseCount] = await sql`SELECT count(*)::int FROM quran_verses`;
    const [translationCount] = await sql`SELECT count(*)::int FROM quran_translations`;
    const [hadithCount] = await sql`SELECT count(*)::int FROM hadiths`;
    const [registryCount] = await sql`SELECT count(*)::int FROM source_registry`;

    console.log("==========================================");
    console.log("📊 Muwāsā Islamic KB Status Report");
    console.log("==========================================");
    console.log(`✅ Source Registry:       ${registryCount.count} registered trusted sources`);
    console.log(`📖 Quran Surahs:          ${surahCount.count} / 114`);
    console.log(`📖 Quran Verses:          ${verseCount.count} verses`);
    console.log(`🌐 Quran Translations:    ${translationCount.count} translations`);
    console.log(`📜 Authentic Hadiths:     ${hadithCount.count} hadiths`);
    console.log("==========================================");

    if (surahCount.count === 114 && verseCount.count >= 6236) {
      console.log("🎉 Core Quran canonical corpus is COMPLETE & VERIFIED.");
    } else {
      console.log("ℹ️ Knowledge base partially seeded. Run ingestion scripts to complete.");
    }
  } catch (err) {
    console.error("❌ Validation query failed:", err);
  } finally {
    await sql.end();
  }
}

validateSources();

File: islamic-kb/scripts/validate_sources.py
import asyncio
from utils import get_db_connection

async def validate_sources():
    conn = await get_db_connection()
    
    print("--- Islamic KB Validation Report ---")
    
    ayah_count = await conn.fetchval("SELECT count(*) FROM quran_verses")
    print(f"Total Ayahs: {ayah_count} (Expected: 6236)")
    
    dupes = await conn.fetchval("SELECT count(*) FROM (SELECT surah_id, ayah_number FROM quran_verses GROUP BY surah_id, ayah_number HAVING count(*) > 1) d")
    print(f"Duplicate Verses: {dupes}")
    
    hadith_count = await conn.fetchval("SELECT count(*) FROM hadiths")
    hadith_no_grade = await conn.fetchval("SELECT count(*) FROM hadiths WHERE grade IS NULL OR grade = ''")
    print(f"Total Hadiths: {hadith_count}")
    print(f"Hadiths missing grade: {hadith_no_grade}")
    
    verses_emb = await conn.fetchval("SELECT count(*) FROM quran_verses WHERE embedding IS NOT NULL")
    hadith_emb = await conn.fetchval("SELECT count(*) FROM hadiths WHERE embedding IS NOT NULL")
    
    if ayah_count > 0:
        print(f"Quran Embedding Coverage: {(verses_emb/ayah_count)*100:.2f}%")
    if hadith_count > 0:
        print(f"Hadith Embedding Coverage: {(hadith_emb/hadith_count)*100:.2f}%")
        
    await conn.close()

if __name__ == "__main__":
    asyncio.run(validate_sources())
